### ISSUE TITLE
hunter_config(Foo GIT_REPOSITORY <URL> GIT_TAG <COMMIT>)

### ISSUE THREAD

**Issue by headupinclouds** [ruslo/hunter#1188]
Is a `hunter_config(Foo GIT_REPOSITORY <URL> GIT_TAG <COMMIT>)` addition (similar to the recent `GIT_SUBMODULE` extension) feasible?  The CMake [ExternalProject](https://cmake.org/cmake/help/v3.0/module/ExternalProject.html#externalproject) docs state that `GIT_TAG` can be a "Git branch name, commit id or tag".  For this to be reproducible, we would probably want to enforce that the specified `GIT_TAG` is a commit id and not a branch name.

**Comment by ruslo** [comment]
I've closed related https://github.com/ruslo/hunter/issues/230 feature request. Would it be easier to add submodule and manage it with Git than manually update `GIT_TAG`?

**Comment by headupinclouds** [comment]
> Would it be easier to add submodule and manage it with Git than manually update GIT_TAG?

My initial sense is that the `GIT_TAG` approach fits some use cases better.

The `GIT_SUBMODULE` approach works well for managing WIP/develop branches via Hunter.  One drawback is that submodules can be confusing for new git users (i.e., `git submodule update --init --recursive`, updates, etc).

The `GIT_TAG` approach seems to be a slightly more succinct option for managing stable upstream dependencies that may or may not have releases or tags.  It also provides a way of creating Hunter packages (i.e., `hunter/cmake/projects/<module>/hunter.cmake`) for projects that don't provide regular `*.tgz` releases.  Submodules seem to lend themselves only to per project `LOCAL` configuration overrides, and don't really seem to fit as a way to create packages directly in Hunter.

**Comment by ruslo** [comment]
> One drawback is that submodules can be confusing for new git users

We are developing on **Git**Hub, is that really a case? :)

> The GIT_TAG approach seems ... that may or may not have releases or tags

If project on GitHub have tags we can use release archive. If project doesn't have tags then you have to use branch name or SHA1 explicitly. With branch name we are losing reproducibility and SHA1 is much more inconvenient than submodule.

> It also provides a way of creating Hunter packages (i.e., hunter/cmake/projects/<module>/hunter.cmake) for projects that don't provide regular *.tgz releases

Such feature was not considered as good originally. You don't have to install Git to use Hunter and release archive downloads faster than `git clone`.

I'm not sure that this feature will be solid. It will make life easier for whom? For users? Windows users have to install Git. MinGW users [will not have access](https://github.com/ingenue/hunter/blob/71bb354732a7cf11b66783a43da5b0ff12e9329d/appveyor.yml#L57-L60) to this feature. Users will pay for `git clone` of repository they are not planning to change. For developers? They will clone repository they don't have access to apply patches.

**Comment by headupinclouds** [comment]
> We are developing on GitHub, is that really a case? :)

I think there are many GitHub users without much git experience beyond `git clone <url>`.  It is a minor point, but in the context of creating a stable `hunter.cmake` package definition, I'd prefer the`GIT_TAG` option over the `GIT_SUBMODULE` approach (I do like the later approach for "under construction" project `config.cmake` overrides).  From the user's perspective (as you point out), it seems `URL` is the best choice where feasible.

I was thinking about this after looking at [boringssl](https://github.com/google/boringssl), which has no tagged releases.  In that case a hunter-packages fork was needed anyway for CMake patches.  All of the CMake patches could be accepted upstream and used directly from Hunter with the `GIT_TAG` feature.  It would be easier to keep the hunter package up to date at that point by simply bumping a `GIT_TAG` field in `cmake/project/BoringSSL/hunter.cmake` for each major upstream change.  I'm just outlining the use case I had in mind here for the discussion.  I see it as more of a feature (similar to the autoconf schemes) that provides another option, and not a replacement for the default URL based package specification.

I also ran into this while reading through tensorflow files, which make use of `ExternalProject_Add` and `GIT_TAG` for some libraries.  I hadn't seen it until then.

> I'm not sure that this feature will be solid. It will make life easier for whom?  For users? Windows users have to install Git. MinGW users will not have access to this feature. Users will pay for git clone of repository they are not planning to change. For developers? They will clone repository they don't have access to apply patches.

Thanks for the discussion and background.  Those are good points, and I have no plans to add such a feature.  If a number of the google projects from the tensorflow effort  (or any upstream repos) accepted CMake patches (mostly package config stuff) without creating releases, I think there would be a stronger case for the `GIT_TAG` feature.

Closing...
